lib/wovnrb/html_replacers/link_replacer.rb

```diff
@@ -13,26 +13,67 @@ def initialize(store, pattern, headers)
       @headers = headers
     end
 
+
     def replace(dom, lang)
+      base_url = base_href(dom)
+
       dom.xpath('//*[match(.)]', MultiTagMatcher.new).each do |node|
         next if wovn_ignore?(node)
 
         href = node.get_attribute('href')
         next if href =~ /^\s*\{\{.+\}\}\s*$/
         next if href =~ /^\s*javascript:/i
         next if is_file?(href)
-        new_href = lang.add_lang_code(href, @pattern, @headers)
+
+        new_href = href
+        new_href = adjust_link_by_base(new_href, base_url) if base_url
+        new_href = lang.add_lang_code(new_href, @pattern, @headers)
+
         node.set_attribute('href', new_href)
       end
     end
 
+    private
+
+    def adjust_link_by_base(href, base_url)
+      return href if href =~ /^\//            # absolute path
+      return href if href =~ /^http(s?):\/\// # full url
+
+      File.join(base_url, href)
+    end
+
     def is_file?(href)
       img_files = /^(https?:\/\/)?.*(\.(#{FileExtension::IMG_FILES}))((\?|#).*)?$/i
       audio_files = /^(https?:\/\/)?.*(\.(#{FileExtension::AUDIO_FILES}))((\?|#).*)?$/i
       video_files = /^(https?:\/\/)?.*(\.(#{FileExtension::VIDEO_FILES}))((\?|#).*)?$/i
       doc_files = /^(https?:\/\/)?.*(\.(#{FileExtension::DOC_FILES}))((\?|#).*)?$/i
       href =~ img_files || href =~ audio_files || href =~ video_files || href =~ doc_files
     end
+
+    def base_href(dom)
+      base_tag = dom.xpath('//base').first
+      return nil unless base_tag
+
+      href = base_tag.get_attribute('href')
+      return href if href =~ /^\//            # absolute path
+      return href if href =~ /^http(s?):\/\// # full url
+
+      abs_path = Addressable::URI.join('/', @headers.dirname, href).to_s
+      strip_relative_path_jumps(abs_path)
+    end
+
+    def strip_relative_path_jumps(path)
+      components = path.split('/') - ['.', '']
+      while components.index('..')
+        if components.index('..').zero?
+          components.shift
+        else
+          components.slice!(components.index('..') - 1, 2)
+        end
+      end
+
+      Addressable::URI.join('/', components.join('/')).to_s
```

Comment:
I meant: can we use `Addressable::URI.join` to replace home made code:
```
      while components.index('..')
        if components.index('..').zero?
          components.shift
        else
          components.slice!(components.index('..') - 1, 2)
        end
      end
```

Maybe something like:
```
components = path.split('/')
Addressable::URI.join(*components).to_s
```

Reply:
ah, I misunderstood your comment. Will check. Thanks

Reply:
Refactor to remove home made code https://github.com/WOVNio/wovnrb/pull/124/commits/946d6db1f1f25367a21f27939dcc5bb2cad7625b